### ISSUE TITLE
Documentation of goals, justification, constraints, and more...

### ISSUE THREAD

**Issue by NachtRitter2** [code4sac/bikecounter#1]
Hi all,

I'll apologize ahead of time if I offend anyone; you can blame it on me being "that noob".

I'm a first time participant in a Code4Sac project and so am not familiar with the approaches used to manage and coordinate the work across the distributed developers.  Looking at this GitHub site, I have no sense of what your timeboxed time frame is (one week, two weeks, four weeks?) or what goals you have per time box (sprint).  If I wanted to participate in the development, I see no way to view the top priority tasks, claim a task, work it and then mark it complete when done.  I do see the "anyone may contribute" message in the README, but no one to contact or no place to go to determine what help is needed.

The second thing I am not seeing which I would expect out of this kind of project is some kind of discussion which explains (or 'justifies') the choices that are being made now.  For example, I see that Marionette is being used for this project... what is the reason behind that?  Is it because someone on the team used it before and is familiar with it and so that makes it the best choice?  Or are the reasons for choosing it well aligned with the needs of the project?  What are the benefits and what are the issues with choosing this approach?  I think of the outcome of this project as not only something that SABA will use for many years to come, but also something that hundreds if not thousands of other local bicycle advocacy groups will want to use for the same purpose to replace the paper forms from NBPD.  So it will mean that other developers will need to work on the code and will need to know not only the how but also the why of the app design.  I suppose one good example of an unknown "why" is: Why is this application called "bike counter" (per the welcome message "Welcome to Bike Counter!") when the scope of work indicated that the application will be used to document "bicycle and **pedestrian** traffic and trends?"

The third thing I am not seeing is any discussion or capture of the constraints or assumptions that the developers need to work with.  For example, will the design currently being developed work in a situation where the volunteer doing the counting has no data connection, or will that be an absolute requirement?  Does the application assume a back end server which each group that uses the app would then need to set up and maintain?  Does the application need to prevent the mobile device from locking the screen so that infrequent data entry is not more difficult than it needs to be?  What is the smallest screen size or smallest memory footprint that can realistically be supported?  And so on.  Some of these constraints can have a significant impact to the design, and can require quite a bit of rework if not considered beforehand.

If the idea per Ash's input is to use this app as an opportunity to document how something is built from beginning to end, then I believe it will have be more comprehensive than just more descriptive commits... though that certainly will be very useful as well!  Some of the documentation I can help out with; for example, by laying out the GitHub wiki with sections to capture things like the project overview, justifications, constraints, and other information.  To capture data about the sprints, etc., it'd probably be best to use a tool designed for that (like JIRA or RedMine) though that obviously adds overhead.

Thoughts and feedback welcome.

Mike

**Comment by jesserosato** [comment]
@NachtRitter2 

Hi Mike,

Thanks for your thorough feedback! Let me try and address your concerns in the order you raised them.
1. Scheduling
   At Code for Sacramento, we do not set delivery deadlines if they can be avoided. We are passionate about providing great software that improves the quality of life in Sacramento, but we are also volunteers with jobs and families and other obligations. We believe the pressure of artificial deadlines increases burnout and lowers our long-term effectiveness. Instead, we believe that a product should be delivered as early and often as possible in order to get user feedback as early and often as possible. At our meetup on Wednesday, we discussed this project with the "client" (SABA representative Jim Brown), and came up with a basic set of requirements, which were documented in some mockup sketches. I will be adding those mockups to the repository this afternoon. When we feel the product satisfies the requirements captured in those mockups, we will release an initial beta. From there, we'll iterate on existing issues and user feedback to improve the product as time allows.
2. Tasks and Task Prioritization
   You're just a little early on this one :smile: We'll soon be adding a series of tasks describing the work that needs to be done. You'll be able to find these tasks by clicking the 'Issues' menu item on the BikeCounter repo - the one you used to create this post :smile: Tasks will not be prioritized, but we will aggressively manage the issues queue to ensure that all issues in the queue at a given time are able to be completed in parallel, and that blocking tasks are completed before their dependencies are added to the queue. We will add readme files titled 'Roadmap' and 'Backlog' that will serve as containers for user stories and tasks (respectively) awaiting entry to the queue. User stories from 'Roadmap' will be decomposed into 'Backlog' tasks, and 'Backlog' tasks will become Github issues as existing and blocking tasks in the issues queue are cleared. We will not use a separate task management tool, as past experience has indicated that such tools are, as you suggest, too much overhead and a barrier to entry for new potential contributors.
3. Decision-making
   At Code for Sacramento, decisions about application architecture, tooling, etc, are generally made via informal discussion during meetups. If you'd like to be involved in these decisions, but aren't able to attend meetups, please let me know and we can video conference you in when we discuss the project. These decisions could certainly stand to be better documented, and we'd love any help you're willing to offer in that department. It's tough to find volunteers who are passionate about documentation! As far as the name, it's mostly just a placeholder. We're more than open to suggestions :smile:
4. Constraints and assumptions
   This is another area that was discussed at the meetup, but could certainly use documentation. I'm happy to talk in more detail about these with you if you're interested in writing them up.
5. Tutorial
   Any help you want to offer in developing educational resources based on this project would be amazing! I believe a tutorial like the one under discussion could easily be its own project, hosted and discussed separately, and we can update documentation and guidelines on this repo as needed. For now, I'll open a new issue on this repo for discussion about the future of the tutorial.

I hope this is at least somewhat helpful. I'm going to close this issue, but please feel free to contact me directly at jrosato at codeforamerica.org if you'd like to talk more. Thanks again for your interest in Code for Sacramento and this project.
- Jesse

**Comment by roughani** [comment]
Thanks for the informative response, @jesserosato! Just piggy-backing on the process documentation topic: we could definitely use some help there. The incorporation of commits is purely to have a paper trail. We can go back and reconstruct the process at a higher, thematic level. I'm particularly interested in capturing video when we test the app with real users to help articulate to show the value of our user-centered, iterative approach to developing applications. This will be super-useful to our government partners. In terms of the specific documentation platform, we'll cross that bridge when we get to it, but it could be as simple as a blog post or page on our website. Hopefully you can help us out with that @NachtRitter2! Thanks for your questions.

**Comment by NachtRitter2** [comment]
@jesserosato 

Thanks Jesse; I appreciate your detailed response. 

BTW, I don't mean to imply "schedule" when I asked about time boxing... but rather was thinking along the lines of a sprint, where the team gets together on a periodic basis and discusses the tasks achieved so far, demonstrates the state of the project, discusses next steps, and assigns (or reassigns) the tasks from the backlog.  Given the makeup of the development team and the constraints you described, it seems particularly important to have a way to track who is working on what and to have a way to transfer ownership from the original owner to a new owner if life "gets in the way" as it often does.  The "one week, two weeks, four weeks" question was related to how often this kind of meeting happens.

I'm looking forward to seeing the mockup sketches and the Roadmap & Backlog readme files.  I think posting those would help quite a bit.

I'm looking forward to the next meetup and will ping you by mail in the meantime to see what (if anything) I can start working on.

~Mike

**Comment by jesserosato** [comment]
Hey Mike,

Sorry for misunderstanding you on the schedule thing. We’ve been meeting once a month during the summer, with meeting dates being posted to the Code for Sac meetup. We don’t formalize sprints or anything, but we do try and establish the work that we’re trying to get done before the next meeting. Mockups, roadmap and backlog are now in the repo in the ‘meta’ directory, but I haven’t had a chance to breakdown anything from the roadmap into tasks yet. Please do shoot me an email, I’d love to talk more about the project with you!
- Jesse

—
